### DISCUSSION TITLE
Stackstorm supported cron expressions

### DISCUSSION THREAD

**Discussion by sarthak1996** [StackStorm/st2#5366] [General]
Hi,
I have been going through stackstorm documentation and I wasn't able to get a proper document detailing which cron expressions are supported by stackstorm. I tried to create a couple of expressions that included '#' but they did not work. Please share a document detailing the list of supported expressions and is there a way to achieve the complete functionality of cron expression if a few of them are not supported.

**Comment by sarthak1996** [comment]
Also i was going through the code, and got to know that APScheduler is being used for cron. I wanted to understand a reason why Stackstorm doesn't support everything that APScheduler's cron class supports. In the crontimer I could see that `additionalProperties:false` has been set which does not allow to pass extra params and hence can not utilize the full capacity of APscheduler.

Is there any other way to achieve this?
I want to use all the APScheduler functionalities. eg. start_date,end_date. Currently I can not pass this as `additionalProperties` is set to False. Although you are just unzipping the dictionary (which ideally should support everything in apscheduler) but since there is a restriction from Stackstorm APIs for schema validation I can not completely use all the params of APscheduler cron

**Comment by arm4b** [comment]
The documentation we currently have is here: https://docs.stackstorm.com/rules.html#core-st2-crontimer with some examples and formats explained.
Yeah, under the hood https://apscheduler.readthedocs.io/en/stable/modules/triggers/cron.html is used.

Per https://github.com/StackStorm/st2/blob/master/st2reactor/requirements.txt#L8 looks like we rely on the current latest `apscheduler==3.7.0`.

**Reply by sarthak1996** [reply]
Yeah so is there a limitation for not supporting all the functionalities of APScheduler cron? Like start_date,end_date,jitter etc.
These parameters can not be passed to crontimer and hence these features are not available

**Reply by arm4b** [reply]
I guess the original idea was to be close to native cron format to make it easy and intuitive for the users.
That doesn't include additional parameters like start_date, end_date, etc.

For the cron format, could you provide a few examples that didn't work?

**Reply by sarthak1996** [reply]
There was an expression with # but I guess I figured it out when I got to know apscheduler cron was used. 

So can this be a feature request. I guess adding start_date and end_date to the params still makes it intuitive to users. I am asking this because we use stackstorm to create and schedule workflows and we are thinking of having this functionality. Currently if we create a rule it starts from that time onwards, but we want to enforce the rule after a particular date till particular date

**Reply by arm4b** [reply]
:+1: Sure, please submit a dedicated feature request in https://github.com/StackStorm/st2/issues for additional params.
Or maybe even opening a Pull Request against the st2 codebase.
